website/addons/googledrive/model.py

```diff
@@ -111,218 +101,170 @@ def get_or_create(cls, node, path):
+    def api(self):
+        """Authenticated ExternalProvider instance"""
+        if self._api is None:
+            self._api = GoogleDriveProvider()
+            self._api.account = self.external_account
+        return self._api
 
     @property
-    def has_auth(self):
-        if self.oauth_settings:
-            return self.oauth_settings.access_token is not None
-        return False
-
-    def fetch_access_token(self, force_refresh=False):
-        if self.oauth_settings:
-            return self.oauth_settings.fetch_access_token(force_refresh)
-        return None
-
-    def clear(self):
-        if self.oauth_settings:
-            self.oauth_settings.revoke_access_token()
-            self.oauth_settings = None
-            self.save()
-
-        for node_settings in self.googledrivenodesettings__authorized:
-            node_settings.deauthorize(Auth(self.owner))
-            node_settings.save()
-
-    def save(self, *args, **kwargs):
-        if self.oauth_settings:
-            self.oauth_settings.save()
-        return super(GoogleDriveUserSettings, self).save(*args, **kwargs)
-
-    def delete(self, save=True):
-        self.clear()
-        super(GoogleDriveUserSettings, self).delete(save)
-
-    def __repr__(self):
-        return u'<GoogleDriveUserSettings(user={self.owner.username!r})>'.format(self=self)
-
-
-class GoogleDriveNodeSettings(StorageAddonBase, AddonNodeSettingsBase):
-
-    folder_id = fields.StringField(default=None)
-    folder_path = fields.StringField(default=None)
-
-    user_settings = fields.ForeignField(
-        'googledriveusersettings', backref='authorized'
-    )
+    def complete(self):
+        return bool(self.has_auth and self.user_settings.verify_oauth_access(
+            node=self.owner,
+            external_account=self.external_account,
+            metadata={'folder': self.drive_folder_id}
+        ))
 
-    @property
     def folder_name(self):
-        if not self.folder_id:
+        if not self.drive_folder_id:
             return None
 
         if self.folder_path != '/':
-            return unquote(os.path.split(self.folder_path)[1].encode('utf-8')).decode('utf-8')
-
-        return '/ (Full Google Drive)'
-
-    @property
-    def has_auth(self):
-        """Whether an access token is associated with this node."""
-        return bool(self.user_settings and self.user_settings.has_auth)
-
-    @property
-    def complete(self):
-        return self.has_auth and self.folder_id is not None
-
-    def deauthorize(self, auth=None, add_log=True):
-        """Remove user authorization from this node and log the event."""
-        if add_log:
-            extra = {'folder': self.folder_name}
-            nodelogger = GoogleDriveNodeLogger(node=self.owner, auth=auth)
-            nodelogger.log(action="node_deauthorized", extra=extra, save=True)
+            return urllib.unquote(os.path.split(self.folder_path)[1].encode('utf-8')).decode('utf-8')
```

Comment:
As discussed earlier, there should be a source code comment to explain the reason for the particular contortions in this line of code.

I'm ok with complexity; I just want to be sure that we understand and document the reasons for it.